src/vm/functions/assets.rs

```diff
@@ -0,0 +1,211 @@
+
+        Ok(Value::okay(Value::Bool(true)))
+    } else {
+        Err(UncheckedError::InvalidArguments("mint-asset! expects a to principal".to_string()).into())
+    }
+}
+
+// Error Codes => 1: from principal is not the current owner of the asset
+//                2: from == to
+//                3: asset does not exist
+pub fn special_transfer_asset(args: &[SymbolicExpression],
+                              env: &mut Environment,
+                              context: &LocalContext) -> Result<Value> {
+    check_argument_count(4, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(UncheckedError::InvalidArgumentExpectedName)?;
+
+    let asset =  eval(&args[1], env, context)?;
+    let from  =  eval(&args[2], env, context)?;
+    let to    =  eval(&args[3], env, context)?;
+
+    let expected_asset_type = env.global_context.database.get_asset_key_type(&env.contract_context.name, asset_name)?;
+
+    if !expected_asset_type.admits(&asset) {
+        return Err(UncheckedError::TypeError(expected_asset_type.to_string(), asset).into())
+    }
+
+    if let (Value::Principal(ref from_principal),
+            Value::Principal(ref to_principal)) = (from, to) {
+
+        if from_principal == to_principal {
+            return Ok(Value::error(Value::Int(2)))
+        }
+
+        let current_owner = match env.global_context.database.get_asset_owner(&env.contract_context.name, asset_name, &asset) {
+            Ok(owner) => Ok(owner),
+            Err(Error::Runtime(RuntimeErrorType::NoSuchAsset, _)) => return Ok(Value::error(Value::Int(3))),
+            Err(e) => Err(e)
+        }?;
+            
+
+        if current_owner != *from_principal {
+            return Ok(Value::error(Value::Int(1)))
+        }
+
+        env.global_context.database.set_asset_owner(&env.contract_context.name, asset_name, &asset, to_principal)?;
+
+        env.global_context.log_asset_transfer(from_principal, &env.contract_context.name, asset_name, 1)?;
+
+        Ok(Value::okay(Value::Bool(true)))
+    } else {
+        Err(UncheckedError::InvalidArguments("transer-asset! expects a from principal and a to principal".to_string()).into())
+    }
+}
+
+// Error Codes => 1: not enough balance
+//                2: from == to
+pub fn special_transfer_token(args: &[SymbolicExpression],
+                              env: &mut Environment,
+                              context: &LocalContext) -> Result<Value> {
+    check_argument_count(4, args)?;
+
+    let token_name = args[0].match_atom()
+        .ok_or(UncheckedError::InvalidArgumentExpectedName)?;
+
+    let amount = eval(&args[1], env, context)?;
+    let from =   eval(&args[2], env, context)?;
+    let to =     eval(&args[3], env, context)?;
+
+    if let (Value::Int(amount),
+            Value::Principal(ref from_principal),
+            Value::Principal(ref to_principal)) = (amount, from, to) {
+        if amount < 0 {
```

Comment:
Should be an `(err X)` if amount == 0?

Reply:
Yeah -- I'm updating it to use an err code here instead of aborting.

Reply:
Addressed in 1cf8cf4